Protect internal coordinator state

elupus opened the pull request:
<!--
  You are amazing! Thanks for contributing to our project!
  Please, DO NOT DELETE ANY TEXT from this template! (unless instructed).
-->
## Breaking change
<!--
  If your PR contains a breaking change for existing users, it is important
  to tell them what breaks, how to make it work again and why we did this.
  This piece of text is published with the release notes, so it helps if you
  write it towards our users, not us.
  Note: Remove this section if this PR is NOT a breaking change.
-->


## Proposed change
<!--
  Describe the big picture of your changes here to communicate to the
  maintainers why we should accept this pull request. If it fixes a bug
  or resolves a feature request, be sure to link to that issue in the
  additional information section.
-->

Update coordinator is triggered from background interval, debouncer and from direct async_refresh() call. All of these can execute the _async_refresh function at the same time causing the state to be broken.

Ensure all paths lock the same async lock to ensure we protect the internal state of the coordinator.

Correct a few more integrations that was missed during the changes to background task in https://github.com/home-assistant/core/pull/112726

## Type of change
<!--
  What type of change does your PR introduce to Home Assistant?
  NOTE: Please, check only 1! box!
  If your PR requires multiple boxes to be checked, you'll most likely need to
  split it into multiple PRs. This makes things easier and faster to code review.
-->

- [ ] Dependency upgrade
- [x] Bugfix (non-breaking change which fixes an issue)
- [ ] New integration (thank you!)
- [ ] New feature (which adds functionality to an existing integration)
- [ ] Deprecation (breaking change to happen in the future)
- [ ] Breaking change (fix/feature causing existing functionality to break)
- [ ] Code quality improvements to existing code or addition of tests

## Additional information
<!--
  Details are important, and help maintainers processing your PR.
  Please be sure to fill out additional details, if applicable.
-->

- This PR fixes or closes issue: fixes #
- This PR is related to issue: https://github.com/home-assistant/core/pull/112726
- Link to documentation pull request: 
- Link to developer documentation pull request: 
- Link to frontend pull request: 

## Checklist
<!--
  Put an `x` in the boxes that apply. You can also fill these out after
  creating the PR. If you're unsure about any of them, don't hesitate to ask.
  We're here to help! This is simply a reminder of what we are going to look
  for before merging your code.

  AI tools are welcome, but contributors are responsible for *fully*
  understanding the code before submitting a PR.
-->

- [x] I understand the code I am submitting and can explain how it works.
- [x] The code change is tested and works locally.
- [x] Local tests pass. **Your PR cannot be merged unless tests pass**
- [x] There is no commented out code in this PR.
- [x] I have followed the [development checklist][dev-checklist]
- [x] I have followed the [perfect PR recommendations][perfect-pr]
- [x] The code has been formatted using Ruff (`ruff format homeassistant tests`)
- [x] Tests have been added to verify that the new code works.
- [x] Any generated code has been carefully reviewed for correctness and compliance with project standards.

If user exposed functionality or configuration variables are added/changed:

- [ ] Documentation added/updated for [www.home-assistant.io][docs-repository]

If the code communicates with devices, web services, or third-party tools:

- [ ] The [manifest file][manifest-docs] has all fields filled out correctly.  
      Updated and included derived files by running: `python3 -m script.hassfest`.
- [ ] New or updated dependencies have been added to `requirements_all.txt`.  
      Updated by running `python3 -m script.gen_requirements_all`.
- [ ] For the updated dependencies - a link to the changelog, or at minimum a diff between library versions is added to the PR description.

<!--
  This project is very active and we have a high turnover of pull requests.

  Unfortunately, the number of incoming pull requests is higher than what our
  reviewers can review and merge so there is a long backlog of pull requests
  waiting for review. You can help here!
  
  By reviewing another pull request, you will help raise the code quality of
  that pull request and the final review will be faster. This way the general
  pace of pull request reviews will go up and your wait time will go down.
  
  When picking a pull request to review, try to choose one that hasn't yet
  been reviewed.

  Thanks for helping out!
-->

To help with the load of incoming pull requests:

- [x] I have reviewed two other [open pull requests][prs] in this repository.

[prs]: https://github.com/home-assistant/core/pulls?q=is%3Aopen+is%3Apr+-author%3A%40me+-draft%3Atrue+-label%3Awaiting-for-upstream+sort%3Acreated-desc+review%3Anone+-status%3Afailure

<!--
  Thank you for contributing <3

  Below, some useful links you could explore:
-->
[dev-checklist]: https://developers.home-assistant.io/docs/development_checklist/
[manifest-docs]: https://developers.home-assistant.io/docs/creating_integration_manifest/
[quality-scale]: https://developers.home-assistant.io/docs/integration_quality_scale_index/
[docs-repository]: https://github.com/home-assistant/home-assistant.io
[perfect-pr]: https://developers.home-assistant.io/docs/review-process/#creating-the-perfect-pr

elupus committed:
Protect internal coordinator state

Update coordinator is triggered from background interval,
debouncer and from direct async_refresh() call. All of these
can execute the async _async_refresh function
at the same time causing the state to be broken.

Ensure all paths lock the same async lock to ensure we
protect the internal state of the coordinator.

elupus commented on homeassistant/helpers/update_coordinator.py:
```diff
@@ -365,7 +376,8 @@ async def _async_setup(self) -> None:
     async def async_refresh(self) -> None:
         """Refresh data and log errors."""
-        await self._async_refresh(log_failures=True)
+        async with self._debounced_refresh.lock:
```
We cant return though. Since there could be two legitamate callers of async_update(). One which would then not get an update.

If current task is already owner (ie recursive lock) could be allowed to return. Which i think it would be in your example there i think.

elupus replied:
We could even raise in that case

bdraco replied:
Right no `return`, just log I think

elupus replied:
I think i got a better solution. Let it keep track of owner and raising on recursive locks.

Ps. I think we have several integrations missing the background tasks waiting when blocking after ticking time. Unsure why we triggered it to be an issue in some integration with these locks thou.

elupus committed:
Raise on recursive locking

elupus committed:
Correct canary update behaviour

Coordinator has already refreshed on init, so no need to refresh again

Coordinator runs it's update in background tasks, so make sure we
wait for those to finish on update.

elupus committed:
Ensure hassio waits for coordinator background tasks

elupus committed:
Protect for exceptions on lock

elupus committed:
Log if lock is held, normally won't happen

elupus committed:
Add testcase

elupus committed:
PSN must wait for coordinator background tasks

elupus committed:
Use named parameters for playstation

elupus commented on homeassistant/helpers/update_coordinator.py:
```diff
@@ -365,7 +376,8 @@ async def _async_setup(self) -> None:
     async def async_refresh(self) -> None:
         """Refresh data and log errors."""
-        await self._async_refresh(log_failures=True)
+        async with self._debounced_refresh.async_lock():
+            await self._async_refresh(log_failures=True)
```
Debouncer can be constructed external to the coordinator. So it would have to be possible to dynamically change its internal lock which also end up weird. 

This was less weird. We always require a debouncer, so letting it be the lock owner is not perfect, but not horrible.

elupus committed:
Use explicit argument

frenck reviewed (approved):
Thanks, @elupus 👍 

../Frenck

elupus commented:
Seems plausible. There was something fishy with those tests which i thought i fixed with the waits for background tasks. But maybe there is more going on in that integration.

I will look at it as soon as i can. But good if some more eyes are on it.

elupus commented:
Small guess here @epenet can't commit right now. But i'm thinking this line is the culprit: https://github.com/home-assistant/core/blob/e3a6c069972cc9a348053f6bfd94cbffded7f448/tests/components/playstation_network/test_init.py#L133

It's set to exactly 1 day which is the same interval the coordinator has. It should be set to something a little bit larger. Not fully sure that is the cause, but it seems possible.

elupus commented:
Let's see if https://github.com/home-assistant/core/pull/154549 fixes it. Did it from web GUI so not tested locally.
